--- a/jaxlib/mosaic/BUILD
+++ b/jaxlib/mosaic/BUILD
@@ -20,7 +20,7 @@ licenses(["notice"])
 package(
     default_applicable_licenses = [],
     default_visibility = [
-        "//jax/experimental:mosaic_users",
+        "//visibility:public",
     ],
 )

--- a/jaxlib/gpu/BUILD
+++ b/jaxlib/gpu/BUILD
@@ -81,6 +81,7 @@ proto_library(

 cc_proto_library(
     name = "triton_cc_proto",
+    visibility = "//visibility:public",
     compatible_with = None,
     deps = [":triton_proto"],
 ) 
