diff --git a/xla/service/collective_permute_key.cc b/xla/service/collective_permute_key.cc
index 320a7cb190..e71802ef47 100644
--- a/xla/service/collective_permute_key.cc
+++ b/xla/service/collective_permute_key.cc
@@ -34,7 +34,17 @@ std::optional<CollectivePermuteKey> GetCollectivePermuteKey(
   }
 
   const auto* cp = Cast<HloCollectivePermuteInstruction>(instruction);
-  return CollectivePermuteKey{cp->source_target_pairs()};
+  std::vector<std::pair<int64_t, int64_t>> source_target_pairs =
+      cp->source_target_pairs();
+  if (instruction->parent()
+          ->parent()
+          ->config()
+          .debug_options()
+          .xla_enable_enzyme_comms_opt()) {
+    // Canonicalize the source-target pairs so that the order does not matter.
+    std::sort(source_target_pairs.begin(), source_target_pairs.end());
+  }
+  return CollectivePermuteKey{source_target_pairs};
 }
 
 }  // namespace xla
diff --git a/xla/service/spmd/custom_call_handler.cc b/xla/service/spmd/custom_call_handler.cc
index 35c574b289..09b6111ca2 100644
--- a/xla/service/spmd/custom_call_handler.cc
+++ b/xla/service/spmd/custom_call_handler.cc
@@ -439,6 +439,11 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiRotate(
       << "No dimension attribute in SPMD multi rotate op";
   int64_t dim = dim_it->second;
 
+  auto bufferize_it = attrs.find("bufferize");
+  TF_RET_CHECK(bufferize_it != attrs.end())
+      << "No bufferize attribute in SPMD multi rotate op";
+  int64_t bufferize = bufferize_it->second;
+
   auto left_amount_it = attrs.find("left_amount");
   TF_RET_CHECK(left_amount_it != attrs.end())
       << "No left_amount attribute in SPMD multi rotate op";
@@ -467,6 +472,8 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiRotate(
       CeilOfRatio(full_size, participating_shards);
   int64_t post_halo_shard_size = post_elements_per_shard;
 
+  int64_t max_start_index = left_amount + right_amount;
+
   // The multi-rotate  operation computes
   //   rotate_left(x, L)
   //   rotate_left(x, L-1)
@@ -479,15 +486,21 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiRotate(
   // This means that we need the first L elements from the right via a
   // halo exchange. Similarly, we need the last R elements from the left via a
   // halo exchange.
-  TF_ASSIGN_OR_RETURN(auto super_shard_and_offset,
-                      ConstructHaloExchangeSuperShard(
-                          hlo->operand(0), dim,
-                          /*left_amount=*/right_amount,
-                          /*right_amount=*/left_amount,
-                          /*handle_last_shard=*/true, post_halo_shard_size));
+  TF_ASSIGN_OR_RETURN(
+      auto super_shard_and_offset,
+      ConstructHaloExchangeSuperShard(hlo->operand(0), dim,
+                                      /*left_amount=*/right_amount,
+                                      /*right_amount=*/left_amount,
+                                      /*handle_last_shard=*/true,
+                                      max_start_index, post_halo_shard_size));
 
   auto [super_shard, shard_offset] = super_shard_and_offset;
 
+  if (bufferize) {
+    super_shard = b_.AddInstruction(HloInstruction::CreateUnary(
+        super_shard->shape(), HloOpcode::kOptimizationBarrier, super_shard));
+  }
+
   auto create_slice = [&](HloInstruction* val, int64_t start, int64_t limit) {
     Shape slice_shape = val->shape();
     slice_shape.set_dimensions(dim, limit - start);
@@ -532,7 +545,7 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiRotate(
 absl::StatusOr<std::pair<HloInstruction*, HloInstruction*>>
 SpmdPartitioningVisitor::ConstructHaloExchangeSuperShard(
     const HloInstruction* input_operand, int64_t dim, int64_t left_amount,
-    int64_t right_amount, bool handle_last_shard,
+    int64_t right_amount, bool handle_last_shard, int64_t max_start_index,
     int64_t post_halo_shard_size) {
   PartitionedHlo input = GetPartitionedHlo(input_operand);
   HloSharding element_sharding = input_operand->sharding();
@@ -579,11 +592,15 @@ SpmdPartitioningVisitor::ConstructHaloExchangeSuperShard(
             return;
           }
           std::vector<int64_t> dst_idx(indices.begin(), indices.end());
-          dst_idx[dim] += 1;
+          dst_idx[dim] += participating_shards - 1;
           dst_idx[dim] %= participating_shards;
+          if (!handle_last_shard && dst_idx[dim] == participating_shards - 1) {
+            return;
+          }
           pairs.emplace_back(device,
                              element_sharding.tile_assignment()(dst_idx));
         });
+    std::sort(pairs.begin(), pairs.end());
 
     Shape slice_shape = local_input->shape();
     slice_shape.set_dimensions(dim, right_amount);
@@ -629,11 +646,15 @@ SpmdPartitioningVisitor::ConstructHaloExchangeSuperShard(
             return;
           }
           std::vector<int64_t> dst_idx(indices.begin(), indices.end());
-          dst_idx[dim] += participating_shards - 1;
+          dst_idx[dim] += 1;
           dst_idx[dim] %= participating_shards;
+          if (!handle_last_shard && dst_idx[dim] == 0) {
+            return;
+          }
           pairs.emplace_back(device,
                              element_sharding.tile_assignment()(dst_idx));
         });
+    std::sort(pairs.begin(), pairs.end());
 
     Shape dynamic_slice_shape = local_input->shape();
     dynamic_slice_shape.set_dimensions(dim, left_amount);
@@ -678,8 +699,6 @@ SpmdPartitioningVisitor::ConstructHaloExchangeSuperShard(
 
   HloInstruction* super_shard = local_input;
 
-  int64_t max_start_index =
-      (post_halo_shard_size - shard_size) * (participating_shards - 1);
   int64_t max_end_index = max_start_index + post_halo_shard_size;
   int64_t result_right_padding =
       max_end_index - left_amount - shard_size - right_amount;
@@ -776,6 +795,11 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiSlice(
       << "No dimension attribute in SPMD multi slice op";
   int64_t dim = dim_it->second;
 
+  auto bufferize_it = int_attrs.find("bufferize");
+  TF_RET_CHECK(bufferize_it != int_attrs.end())
+      << "No bufferize attribute in SPMD multi slice op";
+  int64_t bufferize = bufferize_it->second;
+
   auto amount_it = int_attrs.find("amount");
   TF_RET_CHECK(amount_it != int_attrs.end())
       << "No amount attribute in SPMD multi slice op";
@@ -798,7 +822,7 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiSlice(
   const HloSharding& element_sharding = hlo->sharding().tuple_elements()[0];
 
   // Check that all tuple elements have the same sharding.
-  for (size_t i = 0; i < hlo->sharding().tuple_elements().size(); ++i) {
+  for (int64_t i = 0; i < hlo->sharding().tuple_elements().size(); ++i) {
     TF_RET_CHECK(hlo->sharding().tuple_elements()[i] == element_sharding)
         << "All elements of MultiSlice output must have the same sharding.";
   }
@@ -817,10 +841,12 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiSlice(
 
   int64_t post_elems_per_shard =
       CeilOfRatio(output_shape.dimensions(dim), participating_shards);
-  // Semantics of MultiSlice return amount + 1 results (or specifically, the
-  // given slice, and amount slices after), the original result is handled by
-  // post_elems_per_shard so we need to add amount to it to get the others.
-  int64_t post_halo_shard_size = post_elems_per_shard + amount;
+
+  TF_RET_CHECK(start_index <= sharded_input_size)
+      << "Start index must be less than or equal to sharded input size.";
+
+  TF_RET_CHECK(limit_index >= (participating_shards - 1) * sharded_input_size)
+      << "Limit index must occur on the last shard.";
 
   // Consider a tensor with 4 shards, whose data we care about is 0-D
   // [xyz01][23456][789AB][CDuvm]
@@ -842,20 +868,19 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiSlice(
   // This ensures all the data is available on the local shard, so we can
   // perform all the slices at once.
 
-  int64_t from_right =
-      post_elems_per_shard + amount -
-      std::min(post_elems_per_shard + amount, sharded_input_size - start_index);
+  int64_t from_right = std::max<int64_t>(
+      0, post_elems_per_shard + amount - sharded_input_size + start_index);
+
   int64_t from_left =
-      post_elems_per_shard -
-      std::min(post_elems_per_shard,
-               limit_index - (participating_shards - 1) * sharded_input_size);
-
-  TF_ASSIGN_OR_RETURN(auto super_shard_and_offset,
-                      ConstructHaloExchangeSuperShard(
-                          input_operand, dim, /*left_amount=*/from_left,
-                          /*right_amount=*/from_right,
-                          /*handle_last_shard=*/false, post_halo_shard_size));
-  auto&& [super_shard, shard_offset] = super_shard_and_offset;
+      std::max<int64_t>(0, (participating_shards - 1) *
+                                   (sharded_input_size - post_elems_per_shard) -
+                               start_index);
+
+  // Semantics of MultiSlice return amount + 1 results (or specifically, the
+  // given slice, and amount slices after), the original result is handled by
+  // post_elems_per_shard so we need to add amount to it to get the others.
+  int64_t post_halo_shard_size = post_elems_per_shard;
+
   // We now must actually index into our super shard and get the desired data.
   // We want the data starting at start_index + per_partition_after *
   // shard_offset in the overall previous array. We know that our local shard
@@ -869,13 +894,26 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiSlice(
   // = (start_index + from_left) + (per_partition_after - per_partition_before)
   // * shard_offset
 
+  int64_t max_start_index =
+      (start_index + from_left) +
+      (post_halo_shard_size - sharded_input_size) * (participating_shards - 1);
+
+  TF_ASSIGN_OR_RETURN(
+      auto super_shard_and_offset,
+      ConstructHaloExchangeSuperShard(
+          input_operand, dim, /*left_amount=*/from_left,
+          /*right_amount=*/from_right,
+          /*handle_last_shard=*/false, max_start_index, post_halo_shard_size));
+  auto&& [super_shard, shard_offset] = super_shard_and_offset;
+
   HloInstruction* shard_to_slice;
-  {
+  int64_t additional_offset = 0;
+  if (post_elems_per_shard != sharded_input_size) {
     HloInstruction* zero_offset = b_.AddInstruction(
         HloInstruction::CreateConstant(LiteralUtil::CreateR0<int32_t>(0)));
 
     Shape slice_shape = super_shard->shape();
-    slice_shape.set_dimensions(dim, post_halo_shard_size);
+    slice_shape.set_dimensions(dim, post_elems_per_shard + amount);
 
     HloInstruction* shard_size_change = b_.AddInstruction(
         HloInstruction::CreateConstant(LiteralUtil::CreateR0<int32_t>(
@@ -898,10 +936,19 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiSlice(
 
     std::vector<int64_t> slice_sizes(super_shard->shape().dimensions().begin(),
                                      super_shard->shape().dimensions().end());
-    slice_sizes[dim] = post_halo_shard_size;
+    slice_sizes[dim] = post_elems_per_shard + amount;
 
     shard_to_slice = b_.AddInstruction(HloInstruction::CreateDynamicSlice(
         slice_shape, super_shard, start_indices, slice_sizes));
+  } else {
+    shard_to_slice = super_shard;
+    additional_offset = start_index + from_left;
+  }
+
+  if (bufferize) {
+    shard_to_slice = b_.AddInstruction(HloInstruction::CreateUnary(
+        shard_to_slice->shape(), HloOpcode::kOptimizationBarrier,
+        shard_to_slice));
   }
 
   std::vector<HloInstruction*> sliced_results;
@@ -910,11 +957,11 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiSlice(
     Shape final_slice_shape = shard_to_slice->shape();
     final_slice_shape.set_dimensions(dim, post_elems_per_shard);
     std::vector<int64_t> slice_starts(final_slice_shape.dimensions().size(), 0);
-    slice_starts[dim] = i;
+    slice_starts[dim] = i + additional_offset;
     std::vector<int64_t> slice_limits(
         shard_to_slice->shape().dimensions().begin(),
         shard_to_slice->shape().dimensions().end());
-    slice_limits[dim] = i + post_elems_per_shard;
+    slice_limits[dim] = i + post_elems_per_shard + additional_offset;
     sliced_results.push_back(b_.AddInstruction(HloInstruction::CreateSlice(
         final_slice_shape, shard_to_slice, slice_starts, slice_limits,
         std::vector<int64_t>(final_slice_shape.dimensions().size(), 1))));
@@ -955,10 +1002,13 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_Wrap(
 
   int64_t post_wrap_shard_size =
       CeilOfRatio(hlo->shape().dimensions(dim), participating_shards);
-  TF_ASSIGN_OR_RETURN(auto super_shard_and_offset,
-                      ConstructHaloExchangeSuperShard(
-                          hlo->operand(0), dim, left_amount, right_amount,
-                          /*handle_last_shard=*/true, post_wrap_shard_size));
+  int64_t max_start_index =
+      (post_wrap_shard_size - shard_size) * (participating_shards - 1);
+  TF_ASSIGN_OR_RETURN(
+      auto super_shard_and_offset,
+      ConstructHaloExchangeSuperShard(
+          hlo->operand(0), dim, left_amount, right_amount,
+          /*handle_last_shard=*/true, max_start_index, post_wrap_shard_size));
   auto&& [super_shard, shard_offset] = super_shard_and_offset;
   HloInstruction* zero_offset = b_.AddInstruction(
       HloInstruction::CreateConstant(LiteralUtil::CreateR0<int32_t>(0)));
diff --git a/xla/service/spmd/spmd_partitioner.h b/xla/service/spmd/spmd_partitioner.h
index b0ce8b2bb7..e2296523f5 100644
--- a/xla/service/spmd/spmd_partitioner.h
+++ b/xla/service/spmd/spmd_partitioner.h
@@ -920,6 +920,7 @@ class SpmdPartitioningVisitor : public DfsHloVisitorWithDefault {
   ConstructHaloExchangeSuperShard(const HloInstruction* input_operand,
                                   int64_t dim, int64_t left_amount,
                                   int64_t right_amount, bool handle_last_shard,
+                                  int64_t max_start_index,
                                   int64_t post_halo_shard_size);
 };
 
diff --git a/xla/service/spmd/spmd_partitioner_test.cc b/xla/service/spmd/spmd_partitioner_test.cc
index 18135120a5..b85ff56c95 100644
--- a/xla/service/spmd/spmd_partitioner_test.cc
+++ b/xla/service/spmd/spmd_partitioner_test.cc
@@ -3187,7 +3187,7 @@ ENTRY entry {
   %param0 = f32[12] parameter(0), sharding={devices=[4]<=[4]}
   ROOT %custom-call = (f32[12], f32[12], f32[12]) custom-call(%param0),
     custom_call_target="_SPMDInternalOp_MultiRotate",
-    backend_config="dimension=0,left_amount=1,right_amount=1",
+    backend_config="dimension=0,left_amount=1,right_amount=1,bufferize=0",
     sharding={{devices=[4]<=[4]}, {devices=[4]<=[4]}, {devices=[4]<=[4]}}
 })";
 
@@ -3240,7 +3240,7 @@ ENTRY entry {
   ROOT %custom-call = (f64[1520]{0}, f64[1520]{0}, f64[1520]{0}, f64[1520]{0}) custom-call(%param0),
     custom_call_target="_SPMDInternalOp_MultiSlice",
     sharding={{devices=[2]<=[2]}, {devices=[2]<=[2]}, {devices=[2]<=[2]}, {devices=[2]<=[2]}},
-    backend_config="dimension=0,amount=3,start_indices=[6],limit_indices=[1526],strides=[1]"
+    backend_config="dimension=0,amount=3,start_indices=[6],limit_indices=[1526],strides=[1],bufferize=0"
 })";
 
   TF_ASSERT_OK_AND_ASSIGN(auto module,
@@ -3270,7 +3270,7 @@ ENTRY entry {
   ROOT %custom-call = (f64[1520]{0}, f64[1520]{0}) custom-call(%param0),
     custom_call_target="_SPMDInternalOp_MultiSlice",
     sharding={{devices=[2]<=[2]}, {devices=[2]<=[2]}},
-    backend_config="dimension=0,amount=1,start_indices=[7],limit_indices=[1527],strides=[1]"
+    backend_config="dimension=0,amount=1,start_indices=[7],limit_indices=[1527],strides=[1],bufferize=0"
 })";
 
   TF_ASSERT_OK_AND_ASSIGN(auto module,
@@ -3289,6 +3289,66 @@ ENTRY entry {
                           op::Shape("(f64[760], f64[760])")));
 }
 
+TEST_P(SpmdPartitioningTest, CustomCallMultiSliceRealWorldPaddingBug) {
+  absl::string_view hlo_string = R"(
+HloModule module
+
+ENTRY entry {
+  %param0 = f64[1536] parameter(0), sharding={devices=[2]<=[2]}
+  ROOT %custom-call = (f64[1520]{0}, f64[1520]{0}, f64[1520]{0}) custom-call(%param0),
+    custom_call_target="_SPMDInternalOp_MultiSlice",
+    sharding={{devices=[2]<=[2]}, {devices=[2]<=[2]}, {devices=[2]<=[2]}},
+    backend_config="dimension=0,amount=2,start_indices=[5],limit_indices=[1525],strides=[1],bufferize=0"
+})";
+
+  TF_ASSERT_OK_AND_ASSIGN(auto module,
+                          PartitionComputation(hlo_string, /*num_devices=*/2));
+  int collective_permutes = 0;
+  for (auto* inst : module->entry_computation()->instructions()) {
+    if (inst->opcode() == HloOpcode::kCollectivePermute) {
+      collective_permutes++;
+      for (const auto& pair : inst->source_target_pairs()) {
+        // Device 0 should not fetch from device 1.
+        EXPECT_NE(pair, std::make_pair(1LL, 0LL));
+      }
+    }
+  }
+  EXPECT_EQ(collective_permutes, 1);
+}
+
+TEST_P(SpmdPartitioningTest, CustomCallMultiSlicePaddingRequiresNoLeftHalo) {
+  absl::string_view hlo_string = R"(
+HloModule module
+
+ENTRY entry {
+  %param0 = f64[400] parameter(0), sharding={devices=[4]<=[4]}
+  ROOT %custom-call = (f64[398]{0}, f64[398]{0}) custom-call(%param0),
+    custom_call_target="_SPMDInternalOp_MultiSlice",
+    sharding={{devices=[4]<=[4]}, {devices=[4]<=[4]}},
+    backend_config="dimension=0,amount=1,start_indices=[1],limit_indices=[399],strides=[1],bufferize=0"
+})";
+  // Here we compute a slice of [1, 399) and [2, 400)
+  // As a result, we should only need to fetch right halo.
+
+  TF_ASSERT_OK_AND_ASSIGN(auto module,
+                          PartitionComputation(hlo_string, /*num_devices=*/4));
+  std::cerr << module->ToString() << "\n";
+  int collective_permutes = 0;
+  for (auto* inst : module->entry_computation()->instructions()) {
+    if (inst->opcode() == HloOpcode::kCollectivePermute) {
+      collective_permutes++;
+      for (const auto& pair : inst->source_target_pairs()) {
+        // Dest should be to the left of source.
+        EXPECT_EQ(pair.second, pair.first - 1);
+      }
+    }
+    EXPECT_NE(inst->opcode(), HloOpcode::kDynamicSlice);
+    EXPECT_NE(inst->opcode(), HloOpcode::kPad);
+  }
+
+  EXPECT_EQ(collective_permutes, 1);
+}
+
 TEST_P(SpmdPartitioningTest, PartitionCustomCall) {
   absl::string_view hlo_string = R"(
 HloModule cluster_2013453984438090939__.47
