diff --git a/xla/service/collective_permute_key.cc b/xla/service/collective_permute_key.cc
index 320a7cb190..e71802ef47 100644
--- a/xla/service/collective_permute_key.cc
+++ b/xla/service/collective_permute_key.cc
@@ -34,7 +34,17 @@ std::optional<CollectivePermuteKey> GetCollectivePermuteKey(
   }
 
   const auto* cp = Cast<HloCollectivePermuteInstruction>(instruction);
-  return CollectivePermuteKey{cp->source_target_pairs()};
+  std::vector<std::pair<int64_t, int64_t>> source_target_pairs =
+      cp->source_target_pairs();
+  if (instruction->parent()
+          ->parent()
+          ->config()
+          .debug_options()
+          .xla_enable_enzyme_comms_opt()) {
+    // Canonicalize the source-target pairs so that the order does not matter.
+    std::sort(source_target_pairs.begin(), source_target_pairs.end());
+  }
+  return CollectivePermuteKey{source_target_pairs};
 }
 
 }  // namespace xla
diff --git a/xla/service/spmd/custom_call_handler.cc b/xla/service/spmd/custom_call_handler.cc
index 35c574b289..2b345a4c84 100644
--- a/xla/service/spmd/custom_call_handler.cc
+++ b/xla/service/spmd/custom_call_handler.cc
@@ -580,10 +580,14 @@ SpmdPartitioningVisitor::ConstructHaloExchangeSuperShard(
           }
           std::vector<int64_t> dst_idx(indices.begin(), indices.end());
           dst_idx[dim] += 1;
+          if (!handle_last_shard && dst_idx[dim] == participating_shards - 1) {
+            return;
+          }
           dst_idx[dim] %= participating_shards;
           pairs.emplace_back(device,
                              element_sharding.tile_assignment()(dst_idx));
         });
+    std::sort(pairs.begin(), pairs.end());
 
     Shape slice_shape = local_input->shape();
     slice_shape.set_dimensions(dim, right_amount);
@@ -631,9 +635,13 @@ SpmdPartitioningVisitor::ConstructHaloExchangeSuperShard(
           std::vector<int64_t> dst_idx(indices.begin(), indices.end());
           dst_idx[dim] += participating_shards - 1;
           dst_idx[dim] %= participating_shards;
+          if (!handle_last_shard && dst_idx[dim] == 0) {
+            return;
+          }
           pairs.emplace_back(device,
                              element_sharding.tile_assignment()(dst_idx));
         });
+    std::sort(pairs.begin(), pairs.end());
 
     Shape dynamic_slice_shape = local_input->shape();
     dynamic_slice_shape.set_dimensions(dim, left_amount);
@@ -775,6 +783,11 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiSlice(
   TF_RET_CHECK(dim_it != int_attrs.end())
       << "No dimension attribute in SPMD multi slice op";
   int64_t dim = dim_it->second;
+  
+  auto bufferize_it = int_attrs.find("bufferize");
+  TF_RET_CHECK(bufferize_it != int_attrs.end())
+      << "No dimension attribute in SPMD multi slice op";
+  int64_t bufferize = bufferize_it->second;
 
   auto amount_it = int_attrs.find("amount");
   TF_RET_CHECK(amount_it != int_attrs.end())
@@ -902,6 +915,10 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiSlice(
 
     shard_to_slice = b_.AddInstruction(HloInstruction::CreateDynamicSlice(
         slice_shape, super_shard, start_indices, slice_sizes));
+
+    if (bufferize)
+    shard_to_slice = b_.AddInstruction(
+        HloInstruction::CreateUnary(shard_to_slice->shape(), HloOpcode::kOptimizationBarrier, shard_to_slice));
   }
 
   std::vector<HloInstruction*> sliced_results;
