diff --git a/xla/backends/cpu/codegen/builtin_definition_generator.cc b/xla/backends/cpu/codegen/builtin_definition_generator.cc
index 7da8fc5412..a90c4811ed 100644
--- a/xla/backends/cpu/codegen/builtin_definition_generator.cc
+++ b/xla/backends/cpu/codegen/builtin_definition_generator.cc
@@ -85,9 +85,18 @@ static std::optional<llvm::orc::ExecutorSymbolDef> ResolveBuiltinSymbol(
 //===----------------------------------------------------------------------===//
 
 #if defined(PLATFORM_WINDOWS)
-// This function is used by compiler-generated code on windows, but it's not
-// declared anywhere. The signature does not matter, we just need the address.
-extern "C" void __chkstk(size_t);
+#ifdef __MINGW32__
+#if defined(__i386__)
+#undef _alloca
+extern "C" void _alloca(void);
+#elif defined(__x86_64__)
+extern "C" void ___chkstk_ms(void);
+#else
+extern "C" void __chkstk(void);
+#endif
+#else
+extern "C" void __chkstk(void);
+#endif
 #endif
 
 extern "C" {
@@ -240,8 +249,20 @@ static Registry CreateRegistry() {
 #endif
 
 #if defined(PLATFORM_WINDOWS)
+
+#ifdef __MINGW32__
+#if defined(__i386__)
+  registry["__chkstk"] = SymbolDef(_alloca);
+#elif defined(__x86_64__)
+  registry["__chkstk"] = SymbolDef(___chkstk_ms);
+#else
   registry["__chkstk"] = SymbolDef(__chkstk);
 #endif
+#else
+  registry["__chkstk"] = SymbolDef(__chkstk);
+#endif
+
+#endif
 
 #ifdef MEMORY_SANITIZER
   registry["__msan_unpoison"] = SymbolDef(__msan_unpoison);
