--- a/xla/BUILD
+++ b/xla/BUILD
@@ -29,6 +29,7 @@ package_group(
         "//third_party/py/tpu_graphs/...",
         "//tensorflow/compiler/...",
         "//tensorflow/python/tpu/...",
+        "public",
     ],
 )
 
@@ -1165,6 +1166,7 @@ tf_proto_library(
     srcs = ["autotuning.proto"],
     make_default_target_header_only = True,
     protodeps = ["@tsl//tsl/protobuf:dnn_proto"],
+    visibility = ["//visibility:public"],
 )
 
 cc_library(

--- a/xla/mlir/backends/cpu/transforms/BUILD
+++ b/xla/mlir/backends/cpu/transforms/BUILD
@@ -4,7 +4,7 @@ load("@llvm-project//mlir:tblgen.bzl", "gentbl_cc_library")
 
 package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
-    default_visibility = ["//xla:internal"],
+    default_visibility = ["//xla:friends"],
     licenses = ["notice"],
 )
 
 gentbl_cc_library(

--- a/xla/mlir/memref/BUILD
+++ b/xla/mlir/memref/BUILD
@@ -1,6 +1,7 @@
 package_group(
     name = "friends",
     packages = [
+        "public",
         "//xla/mlir/...",
         # copybara:uncomment_begin(google-only)
         # # TODO(ezhulenev): Clean up dependencies that are leforvers from Autofusion project.

--- a/xla/mlir/math/BUILD
+++ b/xla/mlir/math/BUILD
@@ -1,6 +1,7 @@
 package_group(
     name = "friends",
     packages = [
+        "public",
         "//xla/mlir/...",
         # copybara:uncomment_begin(google-only)
         # # TODO(ezhulenev): Clean up dependencies that are leforvers from Autofusion project.
