--- a/xla/BUILD
+++ b/xla/BUILD
@@ -29,6 +29,7 @@ package_group(
         "//third_party/py/tpu_graphs/...",
         "//tensorflow/compiler/...",
         "//tensorflow/python/tpu/...",
+        "public",
     ],
 )
 
@@ -1165,6 +1166,7 @@ tf_proto_library(
     srcs = ["autotuning.proto"],
     make_default_target_header_only = True,
     protodeps = ["@tsl//tsl/protobuf:dnn_proto"],
+    visibility = ["//visibility:public"],
 )
 
 cc_library(
