diff --git a/third_party/shardy/workspace.bzl b/third_party/shardy/workspace.bzl
index 9628ddb01f..3ef0332b70 100644
--- a/third_party/shardy/workspace.bzl
+++ b/third_party/shardy/workspace.bzl
@@ -3,8 +3,8 @@
 load("//third_party:repo.bzl", "tf_http_archive", "tf_mirror_urls")
 
 def repo():
-    SHARDY_COMMIT = "8f0d576921146c4a9fd800fbb8002f02a332539f"
-    SHARDY_SHA256 = "2f622fabcbd9f5ed1a81a382e2d0db9d64f49a6affa61fe5634b0961f96e8a22"
+    SHARDY_COMMIT = "826f349cd94675698f0e8893997bb435cbb3f751"
+    SHARDY_SHA256 = ""
 
     tf_http_archive(
         name = "shardy",
