diff --git a/xla/service/spmd/custom_call_handler.cc b/xla/service/spmd/custom_call_handler.cc
index cba08934ef..cad211a579 100644
--- a/xla/service/spmd/custom_call_handler.cc
+++ b/xla/service/spmd/custom_call_handler.cc
@@ -133,6 +133,7 @@ ParseOpaqueAsAttributesWithArrays(const HloInstruction* hlo) {
   return std::make_pair(int_attrs, array_attrs);
 }
 
+constexpr char kSPMDOpMultiPad[] = "_SPMDInternalOp_MultiPad";
 constexpr char kSPMDOpRotateRight[] = "_SPMDInternalOp_RotateRight";
 constexpr char kSPMDOpMultiRotate[] = "_SPMDInternalOp_MultiRotate";
 constexpr char kSPMDOpMultiSlice[] = "_SPMDInternalOp_MultiSlice";
@@ -547,7 +548,8 @@ absl::StatusOr<std::pair<HloInstruction*, HloInstruction*>>
 SpmdPartitioningVisitor::ConstructHaloExchangeSuperShard(
     const HloInstruction* input_operand, int64_t dim, int64_t left_amount,
     int64_t right_amount, bool handle_last_shard, int64_t max_start_index,
-    int64_t post_halo_shard_size) {
+    int64_t post_halo_shard_size, HloInstruction* pad_value,
+    bool first_shard_uses_pad_value) {
   PartitionedHlo input = GetPartitionedHlo(input_operand);
   HloSharding element_sharding = input_operand->sharding();
 
@@ -710,8 +712,51 @@ SpmdPartitioningVisitor::ConstructHaloExchangeSuperShard(
   if (divisible_by_participating_shards || !handle_last_shard) {
     std::vector<HloInstruction*> concat_ops;
     if (left_halo) {
+      if (first_shard_uses_pad_value) {
+        TF_RET_CHECK(pad_value != nullptr);
+        HloInstruction* zero_shard_offset = b_.AddInstruction(
+            HloInstruction::CreateConstant(LiteralUtil::CreateR0<int32_t>(0)));
+        HloInstruction* is_first =
+            b_.AddInstruction(HloInstruction::CreateCompare(
+                ShapeUtil::ChangeElementType(shard_offset->shape(), PRED),
+                shard_offset, zero_shard_offset, Comparison::Direction::kEq));
+        HloInstruction* broadcasted_pad_value = b_.AddInstruction(
+            HloInstruction::CreateBroadcast(left_halo->shape(), pad_value, {}));
+        left_halo = b_.AddInstruction(HloInstruction::CreateTernary(
+            left_halo->shape(), HloOpcode::kSelect, is_first,
+            broadcasted_pad_value, left_halo));
+      }
       concat_ops.push_back(left_halo);
     }
+    if (first_shard_uses_pad_value) {
+      TF_RET_CHECK(pad_value != nullptr);
+      size_t leftover_offset =
+          full_pre_wrap_size - (participating_shards - 1) * shard_size;
+      Shape index_shape =
+          ShapeUtil::ChangeElementType(local_input->shape(), S32);
+      HloInstruction* iota =
+          b_.AddInstruction(HloInstruction::CreateIota(index_shape, dim));
+      HloInstruction* limit_index =
+          b_.AddInstruction(HloInstruction::CreateConstant(
+              LiteralUtil::CreateR0<int32_t>(leftover_offset)));
+      HloInstruction* shard_size_const =
+          b_.AddInstruction(HloInstruction::CreateConstant(
+              LiteralUtil::CreateR0<int32_t>(shard_size)));
+      HloInstruction* true_limit_index =
+          b_.AddInstruction(HloInstruction::CreateTernary(
+              limit_index->shape(), HloOpcode::kSelect, is_last, limit_index,
+              shard_size_const));
+      HloInstruction* limit_index_broadcasted = b_.AddInstruction(
+          HloInstruction::CreateBroadcast(index_shape, true_limit_index, {}));
+      HloInstruction* is_less = b_.AddInstruction(HloInstruction::CreateCompare(
+          ShapeUtil::ChangeElementType(index_shape, PRED), iota,
+          limit_index_broadcasted, Comparison::Direction::kLt));
+      HloInstruction* broadcasted_pad_value = b_.AddInstruction(
+          HloInstruction::CreateBroadcast(local_input->shape(), pad_value, {}));
+      local_input = b_.AddInstruction(HloInstruction::CreateTernary(
+          local_input->shape(), HloOpcode::kSelect, is_less, local_input,
+          broadcasted_pad_value));
+    }
     concat_ops.push_back(local_input);
     if (right_halo) {
       concat_ops.push_back(right_halo);
@@ -742,7 +787,8 @@ SpmdPartitioningVisitor::ConstructHaloExchangeSuperShard(
                                            right_amount + result_right_padding);
 
       auto padded_super_shard = b_.AddInstruction(HloInstruction::CreatePad(
-          padded_shape, super_shard, zero, paddingConfig));
+          padded_shape, super_shard, pad_value ? pad_value : zero,
+          paddingConfig));
 
       std::vector<HloInstruction*> start_indices(
           pre_wrap_shape.dimensions().size(), zero_offset);
@@ -778,8 +824,9 @@ SpmdPartitioningVisitor::ConstructHaloExchangeSuperShard(
     Shape padded_shape = super_shard->shape();
     padded_shape.set_dimensions(
         dim, padded_shape.dimensions(dim) + result_right_padding);
-    super_shard = b_.AddInstruction(HloInstruction::CreatePad(
-        padded_shape, super_shard, zero, paddingConfig));
+    super_shard = b_.AddInstruction(
+        HloInstruction::CreatePad(padded_shape, super_shard,
+                                  pad_value ? pad_value : zero, paddingConfig));
   }
 
   return std::make_pair(super_shard, shard_offset);
@@ -974,6 +1021,124 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiSlice(
   return absl::OkStatus();
 }
 
+absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiPad(
+    HloInstruction* hlo) {
+  TF_ASSIGN_OR_RETURN(auto int_attrs, ParseOpaqueAsAttributes(hlo));
+
+  auto dim_it = int_attrs.find("dimension");
+  TF_RET_CHECK(dim_it != int_attrs.end())
+      << "No dimension attribute in SPMD multi pad op";
+  int64_t dim = dim_it->second;
+
+  auto bufferize_it = int_attrs.find("bufferize");
+  TF_RET_CHECK(bufferize_it != int_attrs.end())
+      << "No bufferize attribute in SPMD multi pad op";
+  int64_t bufferize = bufferize_it->second;
+
+  auto amount_it = int_attrs.find("amt");
+  TF_RET_CHECK(amount_it != int_attrs.end())
+      << "No amt attribute in SPMD multi pad op";
+  int64_t amount = amount_it->second;
+
+  const HloInstruction* input_operand = hlo->operand(0);
+  PartitionedHlo input = GetPartitionedHlo(input_operand);
+  TF_RET_CHECK(hlo->sharding().IsTuple()) << "MultiPad output must be a tuple.";
+  const HloSharding& element_sharding = hlo->sharding().tuple_elements()[0];
+
+  // Check that all tuple elements have the same sharding.
+  for (size_t i = 0; i < hlo->sharding().tuple_elements().size(); ++i) {
+    TF_RET_CHECK(hlo->sharding().tuple_elements()[i] == element_sharding)
+        << "All elements of MultiPad output must have the same sharding.";
+  }
+
+  const Shape& output_shape = hlo->shape().tuple_shapes(0);
+
+  const int64_t full_input_size = input_operand->shape().dimensions(dim);
+  const int64_t sharded_input_size = input.hlo()->shape().dimensions(dim);
+
+  const int64_t participating_shards =
+      CeilOfRatio(full_input_size, sharded_input_size);
+
+  // Here to perform all the slices at once, we construct a super shard, and
+  // then perform collective permute to exchange the halo elements, and then
+  // slice the results.
+
+  int64_t post_elems_per_shard =
+      CeilOfRatio(output_shape.dimensions(dim), participating_shards);
+
+  TF_RET_CHECK(amount <= sharded_input_size)
+      << "Amountmust be less than or equal to sharded input size.";
+
+  TF_RET_CHECK(amount <= full_input_size -
+                             (participating_shards - 1) * sharded_input_size)
+      << "Amount must fit within occur on the last shard.";
+
+  // Consider a tensor with 4 shards, whose data we care about is 1-D
+  //
+  // Case 1) The amt causes the shard size to increase:
+  // [0123][4567][89AB][CDEF]
+  // and we want to pad 1 element onto each side.
+  // ->
+  // [.0123][45678][9ABCD][EF___]
+  //
+  // Case 2) The amt does not cause the shard size to increase:
+  // [0123][4567][89AB][CDE_]
+  // and we want to pad 1 element onto each side.
+  // ->
+  // [.012][3456][789A][BCDE]
+  // [0123][4567][89AB][CDE.]
+  //
+  // To begin with we will consider the second case, as it is easier.
+  TF_RET_CHECK(post_elems_per_shard == sharded_input_size)
+      << "Post elems per shard must be equal to sharded input size.";
+
+  int64_t from_right = 0;
+  int64_t from_left = amount;
+
+  int64_t post_halo_shard_size = post_elems_per_shard;
+
+  int64_t max_start_index = amount;
+
+  HloInstruction* pad_value = GetPartitionedHlo(hlo->operand(1)).hlo();
+
+  TF_ASSIGN_OR_RETURN(auto super_shard_and_offset,
+                      ConstructHaloExchangeSuperShard(
+                          input_operand, dim, /*left_amount=*/from_left,
+                          /*right_amount=*/from_right,
+                          /*handle_last_shard=*/false, max_start_index,
+                          post_halo_shard_size, pad_value,
+                          /*first_shard_uses_pad_value=*/true));
+
+  HloInstruction* shard_to_slice = super_shard_and_offset.first;
+
+  if (bufferize) {
+    shard_to_slice = b_.AddInstruction(HloInstruction::CreateUnary(
+        shard_to_slice->shape(), HloOpcode::kOptimizationBarrier,
+        shard_to_slice));
+  }
+
+  std::vector<HloInstruction*> sliced_results;
+  sliced_results.reserve(amount + 1);
+  for (int64_t i = 0; i <= amount; ++i) {
+    Shape final_slice_shape = shard_to_slice->shape();
+    final_slice_shape.set_dimensions(dim, post_elems_per_shard);
+    std::vector<int64_t> slice_starts(final_slice_shape.dimensions().size(), 0);
+    slice_starts[dim] = i;
+    std::vector<int64_t> slice_limits(
+        shard_to_slice->shape().dimensions().begin(),
+        shard_to_slice->shape().dimensions().end());
+    slice_limits[dim] = i + post_elems_per_shard;
+    sliced_results.push_back(b_.AddInstruction(HloInstruction::CreateSlice(
+        final_slice_shape, shard_to_slice, slice_starts, slice_limits,
+        std::vector<int64_t>(final_slice_shape.dimensions().size(), 1))));
+  }
+
+  HloInstruction* result_tuple =
+      b_.AddInstruction(HloInstruction::CreateTuple(sliced_results));
+  SetPartitionedHlo(hlo, result_tuple);
+  return absl::OkStatus();
+}
+
 absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_Wrap(
     HloInstruction* hlo) {
   TF_ASSIGN_OR_RETURN(auto attrs, ParseOpaqueAsAttributes(hlo));
@@ -1090,6 +1255,9 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCall(HloInstruction* hlo) {
   if (hlo->custom_call_target() == kSPMDOpMultiSlice) {
     return HandleCustomCallSPMDInternal_MultiSlice(hlo);
   }
+  if (hlo->custom_call_target() == kSPMDOpMultiPad) {
+    return HandleCustomCallSPMDInternal_MultiPad(hlo);
+  }
   if (hlo->custom_call_target() == kSPMDOpWrap) {
     return HandleCustomCallSPMDInternal_Wrap(hlo);
   }
diff --git a/xla/service/spmd/spmd_partitioner.h b/xla/service/spmd/spmd_partitioner.h
index e7b5090dad..51ebb04406 100644
--- a/xla/service/spmd/spmd_partitioner.h
+++ b/xla/service/spmd/spmd_partitioner.h
@@ -909,6 +909,9 @@ class SpmdPartitioningVisitor : public DfsHloVisitorWithDefault {
 
   // Handlers for specific custom call targets.
   // go/keep-sorted start
+  // multi_pad(x, p, dim, amt) = (pad(x, p, dim, lhs=amt, rhs=0),  ...,
+  //                              pad(x, p, dim, lhs=0,   rhs=amt))
+  absl::Status HandleCustomCallSPMDInternal_MultiPad(HloInstruction* hlo);
   // multi_rotate(x, dim, L, R) = (rotate_left(x, L), ..., rotate_left(x, -R))
   absl::Status HandleCustomCallSPMDInternal_MultiRotate(HloInstruction* hlo);
   // mult_slice(x[idxs...], dim, amt) = (x[idxs...], ..., x[idx+amt...])
@@ -924,7 +927,9 @@ class SpmdPartitioningVisitor : public DfsHloVisitorWithDefault {
                                   int64_t dim, int64_t left_amount,
                                   int64_t right_amount, bool handle_last_shard,
                                   int64_t max_start_index,
-                                  int64_t post_halo_shard_size);
+                                  int64_t post_halo_shard_size,
+                                  HloInstruction* pad_value = nullptr,
+                                  bool first_shard_uses_pad_value = false);
 };
 
 }  // namespace spmd
diff --git a/xla/service/spmd/spmd_partitioner_test.cc b/xla/service/spmd/spmd_partitioner_test.cc
index d9aaf782c7..0de04dec06 100644
--- a/xla/service/spmd/spmd_partitioner_test.cc
+++ b/xla/service/spmd/spmd_partitioner_test.cc
@@ -3298,6 +3298,43 @@ ENTRY entry {
                           op::Shape("(f64[760], f64[760])")));
 }
 
+TEST_P(SpmdPartitioningTest, CustomCallMultiPad) {
+  absl::string_view hlo_string = R"(
+HloModule module
+
+ENTRY entry {
+  %param0 = f64[1519,3056] parameter(0), sharding={devices=[2,1]<=[2]}
+  %pad_val = f64[] constant(0), sharding={replicated}
+  ROOT %custom-call = (f64[1520,3056]{1,0}, f64[1520,3056]{1,0}) custom-call(%param0, %pad_val),
+    custom_call_target="_SPMDInternalOp_MultiPad",
+    sharding={{devices=[2,1]<=[2]}, {devices=[2,1]<=[2]}},
+    backend_config="dimension=0,amt=1,bufferize=0"
+})";
+
+  TF_ASSERT_OK_AND_ASSIGN(auto module,
+                          PartitionComputation(hlo_string, /*num_devices=*/2));
+  VLOG(1) << module->ToString();
+  const auto root = module->entry_computation()->root_instruction();
+
+  auto check_partition_id =
+      op::Reshape(op::DynamicSlice(op::Constant(), op::PartitionId()));
+  auto check_is_first = op::Compare(check_partition_id, op::Constant());
+  auto check_is_last = op::Compare(check_partition_id, op::Constant());
+
+  auto left_halo =
+      op::Select(check_is_first, op::Broadcast(op::Constant()),
+                 op::CollectivePermute(op::Slice(op::Parameter(0))));
+
+  auto local_data = op::Select(
+      op::Lt(op::Iota(), op::Broadcast(op::Select(check_is_last, op::Constant(),
+                                                  op::Constant()))),
+      op::Parameter(0), op::Broadcast(op::Constant()));
+
+  auto concat = op::Concatenate(left_halo, local_data);
+
+  EXPECT_THAT(root, op::Tuple(op::Slice(concat), op::Slice(concat)));
+}
+
 TEST_P(SpmdPartitioningTest, CustomCallMultiSliceRealWorldPaddingBug) {
   absl::string_view hlo_string = R"(
 HloModule module
@@ -3341,7 +3378,6 @@ ENTRY entry {
 
   TF_ASSERT_OK_AND_ASSIGN(auto module,
                           PartitionComputation(hlo_string, /*num_devices=*/4));
-  std::cerr << module->ToString() << "\n";
   int collective_permutes = 0;
   for (auto* inst : module->entry_computation()->instructions()) {
     if (inst->opcode() == HloOpcode::kCollectivePermute) {
