diff --git a/xla/service/spmd/custom_call_handler.cc b/xla/service/spmd/custom_call_handler.cc
index a269b9462b..522107e564 100644
--- a/xla/service/spmd/custom_call_handler.cc
+++ b/xla/service/spmd/custom_call_handler.cc
@@ -911,6 +911,7 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiSlice(
   auto& int_attrs = attrs_pair.first;
   auto& array_attrs = attrs_pair.second;
 
+  std::cerr << " multislice: " << hlo->ToString() << "\n";
   auto dim_it = int_attrs.find("dimension");
   TF_RET_CHECK(dim_it != int_attrs.end())
       << "No dimension attribute in SPMD multi slice op";
@@ -932,13 +933,16 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_MultiSlice(
   int64_t limit_index = limit_indices_it->second.at(dim);
 
   const HloInstruction* input_operand = hlo->operand(0);
+  std::cerr << " multislice operand: " << input_operand->ToString() << "\n";
   PartitionedHlo input = GetPartitionedHlo(input_operand);
   TF_RET_CHECK(hlo->sharding().IsTuple())
       << "MultiSlice output must be a tuple.";
   const HloSharding& element_sharding = hlo->sharding().tuple_elements()[0];
 
+  std::cerr << " elem sharding: " << element_sharding << "\n";
   // Check that all tuple elements have the same sharding.
   for (size_t i = 0; i < hlo->sharding().tuple_elements().size(); ++i) {
+    std::cerr << " res sharding[i=" << i << "]: " << hlo->sharding().tuple_elements()[i] << "\n";
     TF_RET_CHECK(hlo->sharding().tuple_elements()[i] == element_sharding)
         << "All elements of MultiSlice output must have the same sharding.";
   }
@@ -1090,7 +1094,12 @@ absl::Status SpmdPartitioningVisitor::HandleCustomCallSPMDInternal_Wrap(
       << "No right_amount attribute in SPMD wrap op";
   int64_t right_amount = right_amount_it->second;
 
+  std::cerr << " wrap hlo: " << hlo->ToString() << "\n";
   const HloInstruction* input_operand = hlo->operand(0);
+  std::cerr << " + input_operand: " << input_operand->ToString() << "\n";
+  HloSharding element_sharding = input_operand->sharding();
+  std::cerr << " + elem_sharding: " << element_sharding << "\n";
+
   PartitionedHlo input = GetPartitionedHlo(input_operand);
   const Shape& pre_wrap_shape = input_operand->shape();
   const int64_t full_pre_wrap_size = pre_wrap_shape.dimensions(dim);
