diff --git a/third_party/llvm/workspace.bzl b/third_party/llvm/workspace.bzl
--- a/third_party/llvm/workspace.bzl
+++ b/third_party/llvm/workspace.bzl
@@ -5,6 +5,6 @@
 def repo(name):
     """Imports LLVM."""
-    LLVM_COMMIT = "a7c97252ed18839e87796dffd3926086ab62e8fd"
-    LLVM_SHA256 = "49049408bdf9ae162b95b3eacfb8f9af226195e66636d0c9eff9a61d99be54c5"
+    LLVM_COMMIT = "7c17a1d5df47f6d5ec55aefad4333df4afbcb66d"
+    LLVM_SHA256 = ""
 
     http_archive(
