diff --git a/cargo/private/cargo_build_script.bzl b/cargo/private/cargo_build_script.bzl
index 244c9a70..b1fb094f 100644
--- a/cargo/private/cargo_build_script.bzl
+++ b/cargo/private/cargo_build_script.bzl
@@ -323,7 +323,7 @@ def _cargo_build_script_impl(ctx):
         env = env,
         toolchain = None,
         # Set use_default_shell_env so that $PATH is set, as tools like cmake may want to probe $PATH for helper tools.
-        use_default_shell_env = True,
+        use_default_shell_env = False, # https://github.com/bazelbuild/rules_rust/issues/2665
     )
 
     return [
diff --git a/rust/private/rustc.bzl b/rust/private/rustc.bzl
index 68671387..1229d006 100644
--- a/rust/private/rustc.bzl
+++ b/rust/private/rustc.bzl
@@ -1268,6 +1268,9 @@ def rustc_compile_action(
     # this is the final list of env vars
     env.update(env_from_args)
 
+    # sike
+    env["PATH"] = "/usr/bin"
+
     if hasattr(attr, "version") and attr.version != "0.0.0":
         formatted_version = " v{}".format(attr.version)
     else:
